### DISCUSSION TITLE
Support fort Endwar

### DISCUSSION THREAD

**Discussion by Leon-Cantin** [EliotVU/Unreal-Library#82] [General]
Hello, since the other post was getting quite off topic and I got everything running, I though I would create another post to talk about how to add support for Endwar.
I got Ghidra to decompile the uplay version and have been decompiling it some more for a little while now.
I pulled your branch with the endwar feature but it does not seem to work when I open one of the UPK, I am still getting what seems to be the same error. Is there anything else that needs to be done to activate this new feature? If there's more that needs to be done, just point me in the right direction and I'll try to do it.

**Comment by EliotVU** [comment]
That's strange, have you tried running it through a debugger to verify if it DID indeed detect the package as `EndWar`? Try clean solution to be sure too!

As for what can still be done:

* Most changes are guarded by a guessed version (I couldn't find the serialize functions in Ghidra)
* Some UClass features are deserialized incorrectly, so not all classes are fully deserializable.
* There may potentially be some byte-codes that are out of date.
* Support for other `Engine` classes hasn't been verified nor supported (if necessary) i.e. (UTexture, UModel etc)

**Reply by Leon-Cantin** [reply]
I cleaned my solution and rebuilt it but I am still getting the same error. How do I know if it detected the package as "EndWar"? 
Is there something to look for to find the serialize and deserialize functions?

**Reply by Leon-Cantin** [reply]
Well, I just made a big discovery. I tried to see when the game opens the .upk and .u with Process Explorer. Turns out it's never. I renamed all of them and the game works just fine. And that's probably why you were having trouble.

Everything is probably in those proprietary files.

**Reply by EliotVU** [reply]
> I cleaned my solution and rebuilt it but I am still getting the same error. How do I know if it detected the package as "EndWar"? Is there something to look for to find the serialize and deserialize functions?

Step through the debugger at the deserialize function found in `PackageFileSummary`, where it prints `Console.WriteLine("Build:" + stream.Package.Build);`

**Reply by EliotVU** [reply]
> Well, I just made a big discovery. I tried to see when the game opens the .upk and .u with Process Explorer. Turns out it's never. I renamed all of them and the game works just fine. And that's probably why you were having trouble.
> 
> Everything is probably in those proprietary files.

That's very strange!
